app/models/generic_object.rb

```diff
@@ -35,31 +35,50 @@ def property_attributes
     end
   end
 
+  def property_associations
+    properties.select { |k, _| property_association_defined?(k) }.each_with_object({}) do |(k, _), h|
+      h[k] = _property_getter(k)
+    end
+  end
+
   def delete_property(name)
-    properties.delete(name.to_s)
-    save!
+    if !property_attribute_defined?(name) && !property_association_defined?(name)
+      valid_property_names = generic_object_definition.property_attributes.keys + generic_object_definition.property_associations.keys
+      raise "Invalid property [#{name}]: must be one of #{valid_property_names.join(", ")}"
+    end
+
+    _property_getter(name).tap do
+      properties.delete(name.to_s)
+      save!
+    end
   end
 
   def add_to_property_association(name, objs)
     objs = [objs] unless objs.kind_of?(Array)
     name = name.to_s
     properties[name] ||= []
 
-    klass = property_associations[name].constantize
+    klass = generic_object_definition.property_associations[name].constantize
     selected = objs.select { |obj| obj.kind_of?(klass) }
     properties[name] = (properties[name] + selected.pluck(:id)).uniq if selected
-    save
+    return unless properties_changed?
```

Comment:
Do we need to explicitly check for `properties_changed?`?. I think `save!` internally does the check.